src/lib/proxy-server.ts

```diff
@@ -0,0 +1,138 @@
+import http from 'http';
+import { domainToASCII } from 'node:url';
+import httpProxy from 'http-proxy';
+import { SiteServer } from 'src/site-server';
+import { loadUserData } from 'src/storage/user-data';
+
+let proxyServer: http.Server | null = null;
+let isProxyRunning = false;
+
+/**
+ * Gets the site details for a given domain by looking it up in user data and SiteServer
+ */
+async function getSiteByHost( domain: string ): Promise< SiteDetails | null > {
+	try {
+		const userData = await loadUserData();
+		const site = userData.sites.find(
+			( site ) => domainToASCII( site.customDomain ?? '' ) === domainToASCII( domain )
+		);
+		if ( site ) {
+			const server = SiteServer.get( site.id );
+			return server ? server.details : null;
+		}
+
+		return null;
+	} catch ( error ) {
+		console.error( 'Error looking up domain in user data:', error );
+		return null;
+	}
+}
+
+/**
+ * Attempts to start the proxy server on port 80
+ * This requires admin/root privileges
+ */
+export async function startProxyServer(): Promise< boolean > {
+	if ( isProxyRunning ) return true;
+
+	try {
+		return await startDomainProxy();
+	} catch ( error ) {
+		console.error( 'Failed to start proxy server:', error );
+		return false;
+	}
+}
+
+/**
+ * Starts the proxy server for sites with custom domains
+ */
+async function startDomainProxy(): Promise< boolean > {
+	try {
+		// Create proxy with additional options to preserve host header
+		const proxy = httpProxy.createProxyServer();
+
+		proxy.on( 'error', ( err, req, res ) => {
+			if ( res && 'writeHead' in res ) {
```

Comment:
```suggestion
			if ( res instanceof http.ServerResponse ) {
```
Nit, but this seems to clarify the intention better.